### PR TITLE
Upcxx 2023.3 update for 0.19

### PR DESCRIPTION
**Pull request by bonachea**
Cherry-pick of #36629 for 0.19.2

### CODE REVIEW

**Review by PHHargrove — APPROVED** [review]
I have verified the new sha256 hashes for GASNet-EX and UPC++, and I have visually verified that the new and updated logic in upcxx/package.py appears correct to me.

I have not attempted to verify the added hashes for recent oneapi-level-zero versions.

File: var/spack/repos/builtin/packages/upcxx/package.py [hunk 1]
```diff
@@ -16,7 +16,7 @@ def is_CrayXC():
 
 
 def is_CrayEX():
-    return (spack.platforms.host().name == "cray") and (
+    return (spack.platforms.host().name == "linux" or spack.platforms.host().name == "cray") and (
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/upcxx/package.py]: FWIW. You could simplify this as:

```suggestion
    return (spack.platforms.host().name in ["cray", "linux"]) and (
```

**Review by tldahlgren — COMMENTED** [review]
Confirmed the 10 version sha256.  Have a minor suggestion.  Will give @rscohn2 a chance to comment since this PR touches one of his packages.

File: var/spack/repos/builtin/packages/upcxx/package.py (hunk 1 above)
Comment (bonachea) [reply var/spack/repos/builtin/packages/upcxx/package.py]: Hi @tldahlgren, thanks for the suggestion - resolved in 650f15dba5

**Review by tldahlgren — APPROVED** [review]

**Comment by haampie** [comment]
Closing this in favor of #36634

**Comment by rscohn2** [comment]
Level zero changes look good

**Pull request closed without merging**